### ISSUE TITLE
DomainBed result reproducability

### ISSUE THREAD

**Issue by AmirEstiri** [khanrc/swad#26]
Hi,
I have been trying to reproduce the results of your paper in the domainbed benchmark.
From what I understood, your code (train_all.py) only runs for cases where there is only one test domain. Which is not the domainbed benchmark (domainbed runs for all scenarios even multiple test domains and one train domain).
This causes the results of SWAD to be incomparable to other methods implemented on domainbed.
This is the part of code that I was talking about: (line 144)
`args.test_envs = [[te] for te in range(len(dataset))]`

Could you please elaborate on this issue?

Thanks in advance

**Comment by khanrc** [comment]
DomainBed does not run all scenarios. Their runs only additionally include the cases of two test domains for the leave-one-domain-out validation method. Our code does not support the leave-one-domain-out method, so two test domain cases are not required.

**Comment by AmirEstiri** [comment]
Thank you for answering Junbum.
I believe the leave-one-domain-out method is used for validation and finding the optimal hyperparameters. However, for testing, domainbed still runs all possible cases, even multiple test domains, no matter the validation method (leave-one-out/oracle/training).
To make it clearer, when you run the domainbed sweep code, it runs n*(n-1)/2 instances (where n is the number of domains). After the run, it uses different validation methods (leave-one-out/oracle/training) to choose the best model and use that model to test on all n*(n-1)/2 cases and reports the average accuracy as the final performance.
I would appreciate if you could correct me in case I'm wrong.

Thanks,

**Comment by AmirEstiri** [comment]
From what I understood, leave-one-domain out only works for one test domain, but other model selection methods support multiple test domains. (As mentioned [here](https://github.com/facebookresearch/DomainBed/issues/31)).
I think your method only supports leave-one-out.
Please correct me if I'm wrong.

**Comment by khanrc** [comment]
Leave-one-domain-out requires two test domains and the others (oracle and training-validation) use only one test domain, as mentioned the link you referred. Our main target is training-validation model selection.

**Comment by AmirEstiri** [comment]
I appreciate you answering the questions.
We are trying to reproduce the SWAD results for leave-one-out and oracle domain selection.
Do you have any script that we could use for that?
Thank you for your helps

**Comment by khanrc** [comment]
As stated earlier, our results come from the training-validation method; we do not include results from the leave-one-domain-out and oracle selection. What does "reproduce" mean?

p.s. I'm on vacation until 8/7. My response can be delayed.

**Comment by AmirEstiri** [comment]
I understand that your codes only support training-validation method.
However, I was wondering how can we get the results for other method selection techniques?
Could you share the results if you have access to them or help us in modifying SWAD code to get the results for other model selections?

**Comment by khanrc** [comment]
SWAD chooses the average range by training-validation loss in https://github.com/khanrc/swad/blob/main/domainbed/trainer.py#L250. You can use different losses (oracle or leave-one-out loss) here. However, please note that SWAD is designed for training-validation loss; we do not know the performance when using different loss.

**Comment by AmirEstiri** [comment]
Thank you for your clarification!

**Comment by lakpa-tamang9** [comment]
Hi @khanrc, In the readme it is mentioned following, which clearly contradicts from your point [here] (https://github.com/khanrc/swad/issues/26#issuecomment-1645981047), and [here](https://github.com/khanrc/swad/issues/26#issuecomment-1659200329). Can you please elaborate this? Is leave-one-out different from leave-one-domain-out? If yes, can you please shed some light on this?
![image](https://github.com/user-attachments/assets/1075eeba-6ea4-4d0c-8615-cfd608e0f046)

**Comment by khanrc** [comment]
- leave-one-out: _a cross-validation method_ (i.e., evaluation) that leaves one domain as a test domain. Every domain in a dataset is alternately used as a test domain and the results are averaged.
- leave-one-domain-out: _a model selection method_ from DomainBed, which leaves one domain "again" from the train domains for model selection. This is not an evaluation method.

I understand this might be confusing, but they are different.

**Comment by lakpa-tamang9** [comment]
Thank you for clarification.
